### PR TITLE
MBSA-670 fix test runner with stuck intention router

### PR DESCRIPTION
**Pull request by dsvoronin**
test was timed out when first intention was send before observation started

### CODE REVIEW

File: subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt [hunk 1]
```diff
@@ -32,6 +39,11 @@ public class IntentionsRouter(
 
         logger.debug("sending intention with id: $intentionId [$intention]")
 
+        while (!isObserving.get()) {
```
Comment (dsvoronin) [inline subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt]: I see it as an ugly solution, but at least it works.

IDEA test runner with `repeat until failure` gave me stable fails on 2..300 first runs, and 3000+ is green after this fix introduced.

i have ~70 logs: `Intention router is not being observed yet, waiting...` on 3000 runs

Do you have an idea why it is stuck that way? And maybe there is a better solution?

File: subprojects/test-runner/client/src/test/kotlin/com/avito/runner/scheduler/runner/RunnerIntegrationTest.kt [hunk 2]
```diff
@@ -600,7 +603,7 @@ internal class RunnerIntegrationTest {
         val runner = provideRunner(
             targets = targets,
             devicesProvider = devicesProvider,
-            executionTimeout = Duration.ofSeconds(5),
+            executionTimeout = Duration.ofSeconds(1),
```
Comment (dsvoronin) [inline subprojects/test-runner/client/src/test/kotlin/com/avito/runner/scheduler/runner/RunnerIntegrationTest.kt:606]: Conditions are even tighter :)

File: subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt [hunk 3]
```diff
@@ -6,12 +6,16 @@ import com.avito.runner.service.model.intention.Intention
...
 import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.delay
+import java.util.concurrent.atomic.AtomicBoolean
 
 public class IntentionsRouter(
     private val intentionRoutings: MutableMap<String, Channel<Intention>> = mutableMapOf(),
     loggerFactory: LoggerFactory
 ) {
 
+    private val isObserving = AtomicBoolean(false)
```
Comment (dsvoronin) [inline subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt]: i have no idea what i'am doing here. do we even need atomic here? Are coroutines here produces actual threads?

**edit**: it is working fine even on simple boolean

File: subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt (hunk 1 above)
Comment (sergioboi) [reply subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt]: I think it's an expected behavior that sending happens before observing
The key problems there maybe in that `intentionRoutings` is simple mutable map.
Mutable map can't correctly work in multithreaded environment.

You could try to make `send` and `observe` synchronized

Comment (dsvoronin) [reply subprojects/test-runner/service/src/main/kotlin/com/avito/runner/service/IntentionsRouter.kt]: Yes, you're right!
I used Mutex to synchronize access to map for all methods and it works fine that way

**Review by eugene-krivobokov — APPROVED** [review]

**Pull request merged**